### PR TITLE
sparse complex: add linear algebra, addmm

### PR DESCRIPTION
**Pull request by aocsa**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55739 sparse complex: add linear algebra, addmm**

- [x]  complex support to sparse.addmm  fordward (CPU/CUDA)
- [x]  complex support to sparse.addmm  backward (CPU/CUDA)
- [x]  unit tests

Differential Revision: [D27765617](https://our.internmc.facebook.com/intern/diff/D27765617)

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -289,6 +289,7 @@
 - func: conj.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
   dispatch:
     CPU, CUDA: conj_out
+    SparseCPU, SparseCUDA: conj_out_sparse
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:292]: bloop. conjugate views gonna be hard on sparse but I guess we will figure something out. https://github.com/pytorch/pytorch/pull/44799

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 2]
```diff
@@ -1623,4 +1622,23 @@ Tensor& bmm_out_sparse_cpu(const SparseTensor& self, const Tensor& mat2, Tensor&
...
+Tensor& conj_out_sparse(const Tensor& input, Tensor& result) {
+  TORCH_INTERNAL_ASSERT(input.is_sparse());
+  if (input.numel() == 0) {
+    return result;
+  }
+  if (!is_same_tensor(result, input)) {
+    copy_sparse_to_sparse_(result, input);
+  }
+  Tensor conj_values = at::conj(input._values());
+  result._values().copy_(conj_values);
```
Comment (ezyang) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp:1639]: This seems unnecessary. Just `conj_out` directly in to the result values.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 3]
```diff
@@ -1623,4 +1622,23 @@ Tensor& bmm_out_sparse_cpu(const SparseTensor& self, const Tensor& mat2, Tensor&
   return result;
 }
 
+Tensor conj_sparse(const Tensor& input) {
```
Comment (ezyang) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp:1625]: You need a fast path here for when the tensor in question is not complex, similar to the fast path in the dense conjugate case.

**Comment by ezyang** [comment]
This all seems fine. Need to fix the build failures

**Comment by ezyang** [comment]
Need to ifdef this out for ROCm: 

```
02:21:42 [ 84%] Building HIPCC object caffe2/CMakeFiles/torch_hip.dir/__/aten/src/THH/generated/torch_hip_generated_THHTensorMathReduceChar.hip.o
02:21:43 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseHIPBlas.hip:338:26: error: no matching function for call to 'hipsparseCcsrmm2'
02:21:43   TORCH_CUDASPARSE_CHECK(hipsparseCcsrmm2(handle, opa, opb, i_m, i_n, i_k, i_nnz, alpha, desc, csrvala, csrrowptra, csrcolinda, b, i_ldb, beta, c, i_ldc));
02:21:43                          ^~~~~~~~~~~~~~~~
02:21:43 /var/lib/jenkins/workspace/aten/src/ATen/hip/Exceptions.h:59:31: note: expanded from macro 'TORCH_CUDASPARSE_CHECK'
02:21:43     hipsparseStatus_t __err = EXPR;                              \
02:21:43                               ^~~~
02:21:43 /opt/rocm/hipsparse/include/hipsparse.h:1358:19: note: candidate function not viable: no known conversion from 'rocblas_float_complex *' (aka 'rocblas_complex_num<float> *') to 'const hipComplex *' (aka 'const HIP_vector_type<float, 2> *') for 8th argument
02:21:43 hipsparseStatus_t hipsparseCcsrmm2(hipsparseHandle_t         handle,
02:21:43                   ^
```

**Comment by aocsa** [comment]
For simplicity, I'm opening a new PR, https://github.com/pytorch/pytorch/pull/57129, after base PR merging  failed.
 https://github.com/pytorch/pytorch/pull/54153#issuecomment-827997616

**Pull request closed without merging**